### PR TITLE
Customing Large File Upload Task

### PR DESCRIPTION
**Pull request by muthurathinam**
## Summary

Providing to pass custom options to large file upload task and also a way to upload file to share point / onedrive for business etc

## Motivation

User ask in this thread: https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/111

## Test plan

`npm tun test`

## Types of changes

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

-   [x] I have read the **CONTRIBUTING** document.
-   [x] My code follows the code style of this project.
-   [x] My change requires a change to the documentation.
-   [x] I have updated the documentation accordingly.
-   [ ] I have added tests to cover my changes.
-   [x] All new and existing tests passed.

### CODE REVIEW

File: src/tasks/OneDriveLargeFileUploadTask.ts [hunk 1]
```diff
@@ -39,6 +39,29 @@ export class OneDriveLargeFileUploadTask extends LargeFileUploadTask {
...
+		if (path === "") {
+			path = "/";
+		}
+		if (path[0] !== "/") {
+			path = `/${path}`;
+		}
+		if (path[path.length - 1] !== "/") {
+			path = `${path}/`;
+		}
+		return encodeURI(`/me/drive/root:${path}${fileName}:/createUploadSession`);
```
Comment (MIchaelMainer) [inline src/tasks/OneDriveLargeFileUploadTask.ts]: There are many entitysets that can be targeted for uploading a large item. Per docs, at least /drives. /groups, /sites, /users in addition to the /me singleton. We should support all of them. Additionally, we should support not just uploading to the drive root, but to any arbitrary folder specified by itemid.

File: src/tasks/LargeFileUploadTask.ts [hunk 2]
```diff
@@ -97,6 +97,28 @@ export class LargeFileUploadTask {
 	 */
 	protected nextRange: Range;
 
+	/**
+	 * @public
+	 * @static
+	 * @async
+	 * Makes request to the server to create an upload session
+	 * @param {Client} client - The GraphClient instance
+	 * @param {any} payload - The payload that needs to be sent
```
[the remark this answers is not in the document]
Comment (muthurathinam) [reply src/tasks/LargeFileUploadTask.ts]: yes it contains the meta data that needs to be sent while uploading.

File: src/tasks/LargeFileUploadTask.ts [hunk 3]
```diff
@@ -97,6 +97,28 @@ export class LargeFileUploadTask {
...
+	 * @returns The promise that resolves to LargeFileUploadSession
+	 */
+	public static async createUploadSession(client: Client, requestUrl: string, payload: any): Promise<any> {
```
[the remark this answers is not in the document]
Comment (muthurathinam) [reply src/tasks/LargeFileUploadTask.ts]: * This is a generic class which handles the large file upload, so the url might vary for say workload. yes it is createUploadSession for onedrive but could be something else for exchange if they implement. so this is the reason I didn't put any validation for url here
* Will add parameter for if match option in onedrivelargefileupload

File: src/tasks/OneDriveLargeFileUploadTask.ts (hunk 1 above)
Comment (muthurathinam) [reply src/tasks/OneDriveLargeFileUploadTask.ts]: Yes, this is one of the ask from the user and we are helping to achieve the same using large file upload base class. This is what this PR is all about, previously creating upload session with largeFileUpload base class was not possible, this PR adds that.

**Pull request merged**